### PR TITLE
fix(stash): preserve fixer tail-line deletions in three-way merge

### PR DESCRIPTION
**Pull request by jdx**
## Summary

- Fixes [#929](https://github.com/jdx/hk/discussions/929): fixer changes that remove trailing lines were silently undone when the worktree had an unrelated unstaged change in the middle of the same file.
- Root cause in `src/merge.rs::diff_hunks`: when the LCS walk consumed `other` entirely after a matching line, `cur_start` was `None` and `j == m`, so neither post-loop branch fired and the deletion of `base[i..n]` was dropped. `three_way_merge_hunks` then copied the removed lines back in verbatim.
- Fix: emit an empty-lines deletion hunk for `base[i..n]` when the post-loop state indicates a pure tail deletion.

## Test plan

- [x] New unit test `diff_emits_pure_tail_deletion_hunk` directly asserts the previously-missing hunk.
- [x] New unit test `merge_preserves_fixer_tail_deletion_with_middle_worktree_change` covers the end-to-end three-way merge from the discussion.
- [x] New bats test `test/fixer_tail_deletion_with_middle_unstaged.bats` reproduces the exact scenario from the discussion through a real `git commit` and verifies both worktree and HEAD content.
- [x] `cargo test` — all 158 Rust tests pass.
- [x] Re-ran related newline/stash/EOF bats tests (`stash_default`, `fix_preserves_unstaged_newline`, `pre_commit_preserves_eof_newline`, `newline_stripping_bug`, `pre_commit_prettier_commits_fixer_changes`, `pre_commit_prettier_restaging`) — all 14 pass.
- [x] Manually reproduced with the older `hk 1.44.1` binary to confirm the bug exists pre-fix, then re-ran with the patched binary to confirm the fix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Modifies core diff/merge behavior used during stash-based pre-commit fixes; mistakes could drop or misapply edits in edge cases. Added regression tests reduce risk but the change affects merge correctness across file-ending scenarios.
> 
> **Overview**
> Fixes `src/merge.rs::diff_hunks` to emit a **pure tail-deletion** hunk (empty `lines` over `base[i..n]`) when `other` ends after an LCS match, preventing trailing-line removals from being silently lost during `three_way_merge_hunks`.
> 
> Adds targeted regression coverage: new Rust unit tests for tail-deletion cases and an end-to-end `bats` test reproducing the stash-based pre-commit scenario where a fixer strips trailing blank lines while the worktree has a middle-of-file unstaged change.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e24e25b573681278ea5370c41eee25b8812b1bfa. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/merge.rs [hunk 1]
```diff
@@ -86,6 +86,15 @@ pub fn diff_hunks(base: &str, other: &str, source: HunkSource) -> Vec<Hunk> {
...
+    } else if i < n {
+        // pure tail deletion: other is shorter than base, loop exited with j == m
+        // after a matching line reset cur_start to None
+        hunks.push(Hunk {
+            start: i,
+            end: n,
+            lines: vec![],
+            source,
+        });
     }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/merge.rs:98]: Analyzed this — the scenario `cur_start = Some` at exit with `j == m && i < n` is unreachable. `j` advances only via match (which clears `cur_start`) or insert; insert at `j == m-1` is impossible because `dp[i][m] = 0` (boundary) and `dp[i+1][m-1] >= 0`, so the algorithm always picks delete there. Therefore `cur_start = Some` at exit ⟹ `i == n`, and `end: i` correctly equals `n`. Added unit test `diff_handles_tail_deletion_following_mismatch` in e24e25b that pins this invariant for the exact shape (base `a\nb\nc\nd\n`, other `a\nX\n`).

File: test/fixer_tail_deletion_with_middle_unstaged.bats [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    run bash -lc 'git -c commit.gpgsign=false commit -m "redis"'
+    echo "$output"
+    assert_success
+
+    # After the commit: fixer's tail deletion is preserved on disk and the
+    # worktree retains the version=2.0 unstaged change.
+    expected=$'name: my-app\nversion: 2.0\ndeps:\n  flask: 2.0\n  redis: 4.0\n'
+    actual="$(cat config.yml)"$'\n'
+    [ "$actual" = "$expected" ]
```
[the remark this answers is not in the document]
Comment (jdx) [reply test/fixer_tail_deletion_with_middle_unstaged.bats]: Good catch — fixed in e24e25b. Both worktree and HEAD assertions now use `diff` against a `printf '%s'` of the expected string, so trailing newlines are compared byte-for-byte.

File: test/fixer_tail_deletion_with_middle_unstaged.bats [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    # The committed (HEAD) content is the fixer's output: redis added, no
+    # trailing blank lines.
+    expected_head=$'name: my-app\nversion: 1.0\ndeps:\n  flask: 2.0\n  redis: 4.0\n'
+    actual_head="$(git show HEAD:config.yml)"$'\n'
+    [ "$actual_head" = "$expected_head" ]
```
[the remark this answers is not in the document]
Comment (jdx) [reply test/fixer_tail_deletion_with_middle_unstaged.bats]: Fixed in e24e25b alongside the other assertion — now uses `git show HEAD:config.yml | diff - <(printf '%s' "$expected_head")`.

File: test/fixer_tail_deletion_with_middle_unstaged.bats (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdx) [reply test/fixer_tail_deletion_with_middle_unstaged.bats]: Same issue Gemini flagged on the adjacent lines — already fixed in e24e25b by switching both assertions to byte-exact `diff` comparison.
